### ISSUE TITLE
5.3.1 has too much going on and should be shortened

### ISSUE THREAD

**Issue by tghosth** [OWASP/ASVS#1589]
[5.3.1](https://appsecg.host/5.3) | Verify that output encoding is relevant for the interpreter and context required. For example, use encoders specifically for HTML values, HTML attributes, JavaScript, URL parameters, HTTP headers, SMTP, and others as the context requires, especially from untrusted inputs (e.g. names with Unicode or apostrophes, such as ねこ or O'Hara). (C4)
-- | --

**Comment by tghosth** [comment]
Point from @ImanSharaf https://github.com/OWASP/ASVS/issues/1561#issuecomment-1474479240

> Thank you for your response and for pointing out that ASVS 5.3.1 does technically cover CRLF Injection in HTTP headers. However, I would like to express my concern that the current wording of 5.3.1 might not explicitly highlight the importance of testing for CRLF Injection attacks. As a result, testers might inadvertently overlook this specific attack vector when performing security assessments.

**Comment by elarlang** [comment]
We also should address CSS related issue with this: https://github.com/OWASP/ASVS/issues/1558

**Comment by elarlang** [comment]
As the Pandora box is opened...

We still need to have abstract requirement to cover all not listed syntaxes:
> Verify that output encoding is relevant for the interpreter and context required. 


Separate requirements:
* HTML values, HTML attributes
  * can we add here XML values and XML attributes and split up XMl encoding and XPath sanitize from https://github.com/OWASP/ASVS/issues/1556
* JavaScript - for JavaScript is available escaping or encoding, both should be covered
* URL encoding
* HTTP headers
  * Should it be separate requirement or also URL encoding?
  * Should it cover CRLF injection or it should be also separate requirement?
  * Should it cover converting cookies (missing at the moment)?
  * Should it cover encoding filename for content-disposition field (https://github.com/OWASP/ASVS/issues/1390)
* SMTP - is there such thing like "SMTP encoding?" or we actually should name it Content-Transfer-Encoding?

I did not cover from 5.3.1 - do we need to mention it?
> especially from untrusted inputs (e.g. names with Unicode or apostrophes, such as ねこ or O'Hara).

**Comment by jmanico** [comment]
Note: CSS encoding is a thing, it's a native JS function. CSS.escape for example.

**Comment by tghosth** [comment]
I agree that we should split out this requirement as @elarlang has set out above, the treatment should either be **one or more** of the following:

- Create new requirement if the particular context does not already have a requirement 
- Reference this issue and what is needed if there is already a related open issue (such as with _HTML values, HTML attributes_)
- Make sure that each requirement (existing or new) correctly captures the fix that is needed (encoding, escaping, sanitizing, sandboxing, etc)
- (Don't move existing requirements around yet)
- Make sure that an individual requirement is not mixing different treatments or contexts, e.g. we may want to split out SMTP and IMAP.

**Comment by elarlang** [comment]
For brainstorming - to cover entire chapters 5.2 and 5.3, it's enough to have a requirement:
> Verify that the application builds commands and documents without data used and can not change the command and document structure by using relevant defense for the syntax and context required, such as parametrization, encoding, escaping, or sanitization.

Where we can put the line between "too detailed" vs "too abstract".

**Comment by jmanico** [comment]
> CSS - (for some reason it is missing from issue description but actually exists in bleeding edge version) - As there is no such thing like CSS encoding, we need to use input validation and/or sanitize when using userinput in CSS. Should be covered via #1558

CSS encoding does exist in native javascript and other languages meant to escape CSS Variables. See: https://developer.mozilla.org/en-US/docs/Web/API/CSS/escape_static

**Comment by elarlang** [comment]
Jim, you already said that (https://github.com/OWASP/ASVS/issues/1589#issuecomment-1497956789). Just please, (re)read and learn the issues before commenting :)

**Comment by jmanico** [comment]
Easy there Elar.

**Comment by jmanico** [comment]
> Jim, you already said that ([#1589 (comment)](https://github.com/OWASP/ASVS/issues/1589#issuecomment-1497956789)). Just please, (re)read and learn the issues before commenting :)

Understood and will do.

**Comment by elarlang** [comment]
My a bit updated proposal:

> Verify that output encoding is relevant for the interpreter and context required, such as encoding value for HTML value, HTML attribute, JavaScript, URL parameter, HTTP header, or SMTP.

**Comment by jmanico** [comment]
I like it Elar.

**Comment by tghosth** [comment]
So I am not sure I completely understood your draft, does my rephrasing make sense:

> Verify that output encoding is relevant for the interpreter and context required, such as encoding **the relevant characters** for HTML **elements**, HTML attribute, JavaScript, URL parameter, HTTP header, or SMTP.

**Comment by elarlang** [comment]
It makes again the requirement longer, but should we mention also the goal for the requirement? That it is not possible to manipulate the output command or document structure?

**Comment by tghosth** [comment]
I think we can probably explain the goal in the section text intro. I think trying to include here is a little too complicated.

**Comment by tghosth** [comment]
There is some singular/plural inconsistency which I didn't pick up on before and I will fix.

Do we still want to include SMTP here? Do we mention SMTP anywhere else that we could rely on instead as it feels a little out of place.

**Comment by elarlang** [comment]
> Do we still want to include SMTP here? Do we mention SMTP anywhere else that we could rely on instead as it feels a little out of place.

I prefer to not include SMTP here. The point is to cover principles, not to list all possible encodings in the world, otherwise, we are soon back to the problem that we tried to solve first.

**Comment by tghosth** [comment]
Do we mention SMTP somewhere else that we can rely on?

**Comment by jmanico** [comment]
Output encoding and validating a SMTP url for a UI to stop XSS is a different control than encoding SMTP for a email server. One stops XSS, another stops email injection.

**Comment by elarlang** [comment]
> Output encoding and validating a SMTP url for a UI to stop XSS is a different control than encoding SMTP for a email server. One stops XSS, another stops email injection.

Whatever the syntax or interpretator, the generated command or document must not be manipulatable.

See my comment here: https://github.com/OWASP/ASVS/issues/1589#issuecomment-2068586392

**Comment by jmanico** [comment]
Elar, when addressing different attacks (e.g., XSS vs. Email Server Injection) that require different controls (e.g., Output UI encoding vs. proper use of email server APIs and email server encoding), I suggest we have different requirements.

So I personally like how things are currently in both 5.2.3 (SMTP/IMAP injection) and 5.3.1 (XSS Defense). This level of detail in 5.3.1 is helpful, IMO.

**Comment by jmanico** [comment]
All I am saying is that I like the detail of 5.3.1

**Comment by tghosth** [comment]
> Output encoding and validating a SMTP url for a UI to stop XSS is a different control than encoding SMTP for a email server. One stops XSS, another stops email injection.

@jmanico I think you are referring about `mailto:` links in an HTML file, is that correct?

**Comment by tghosth** [comment]
@jmanico if you are talking about `mailto:` links, do you think that is covered as "URL parameters" in the suggestion below?

> [MODIFIED] Verify that output encoding is relevant for the interpreter and context required, such as encoding the relevant characters for HTML elements, HTML attributes, JavaScript, URL parameters and HTTP headers.

**Comment by jmanico** [comment]
Even an encoded URL still fires XSS if your validation is weak and does not validate the scheme. JavaScript URL’s even when encoded cause XSS in the attribute context. This is why url validation is more crucial than encoding when it comes to untrusted URl’s.
 
This is less a problem of mailto  since typically the dev’s do:

URL = “mailto:” + untrustedemail.

Does this help and make sense who encoding is not encourage for URL’s?

**Comment by elarlang** [comment]
With encoding you just make sure that you can not manipulate the document or command structure.

URL, when displayed in an HTML document (e.g. as `href` or `src` attribute value) requires validation, sanitization and URL encoding. It is all a different requirement and a different topic.

**Comment by jmanico** [comment]
*Sanitization is not required and URL encoding is not the right encoding 
*when putting untrusted URL's in HTML documents.

When placing a untrusted URL in the content of a HTML attribute you only 
need 2 controls.

1) Validation to ensure the URL is a legal URL and is not a Javascript: 
Data: or other dangerous URL.

2) Encoding in the content of use. If you put a URL in a src or href 
attribute then you want to do HTML attribute/HTML Entity encoding, not 
URL encoding.

I think we need a clear requirement for this. Untrusted URL's causing 
XSS is a big problem.

I have mentioned this several times Elar, and you continue to state this 
defense incorrectly. I mean that with respect, I know this is challenging.

- Jim

**Comment by elarlang** [comment]
> URL, when displayed in an HTML document (e.g. as href or src attribute value) requires validation, sanitization and URL encoding. It is all a different requirement and a different topic.

And when I say in first order, you need validation, you just ignore it?

> I think we need a clear requirement for this. Untrusted URL's causing. XSS is a big problem.

This IS separate topic. _I have said it several times_, and you continue to talk about it.

We have requirement 5.1.5 for that.

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **5.1.5** | [MODIFIED, SPLIT TO 50.7.1] Verify that the application will only automatically redirect the user to a different URL directly from an application URL where the destination appears on an allow list. | ✓ | ✓ | ✓ | 601 |

The topic here is 5.3.1. It is for encoding in general.

**Comment by elarlang** [comment]
And technically, it is not correct (in the context of building URL dynamically by the application):

> 2) Encoding in the content of use. If you put a URL in a src or href
attribute then you want to do HTML attribute/HTML Entity encoding, not
URL encoding.

When you build a URL, it must be in the correct URL encoding anyway. When displaying it in a `src` or `href` attribute value, you need to convert the value for HTML **additionally**. If you had the URL in the correct URL encoding, you have only the `&` sign to convert.

**Comment by jmanico** [comment]
Elar, 5.1.5 does not address my concern at all. Allow-list validation is 
a good control but it's not always acceptable for applications handling 
untrusted URL's.

For example, a social media page may ask you to enter in your homepage 
as part of your user profile that is rendered in a anchor tag's href 
attribute. In that case, we would need to validate that the URL is 
valid, and the protocol is only HTTPS. Allow-list validation will not 
work for those kind of use cases.

And then when you add the URL to a HTML attribute, you need to HTML 
entity encode.

Again, 5.1.5 does not cover this use case.

And even though we are discussing encoding, my point is that encoding of 
URL's in a HTML attribute content is not a sufficient control to stop XSS.

I keep bringing this up because I am politely concerned how you are 
discussing this and feel you are missing the point of how proper defense 
of untrusted URL's needs to be handled. And this workflow when done 
wrong causes 1/3 of real world XSS from Googles research.

**Comment by jmanico** [comment]
Building a URL dynamically is an entirely different issue. If you put untrusted data in the PATH of a URL you need to base64url encode like the JWT standard suggests. URL encoding is only safe for URL construction when placing data in a querystring parameter.

**Comment by elarlang** [comment]
Why you address the URL topic here as the problem of the requirement 5.3.1?

> And even though we are discussing encoding, my point is that encoding of URL's in a HTML attribute content is not a sufficient control to stop XSS.

We are not covering "all the possible XSS here". Already said that, many times. If you have a specific problem of handling URLs. please open a separate issue and stay on topic here.

**Comment by jmanico** [comment]
I am absolutely on topic here. Please contribute to the conversation positively or hold your comments.

**Comment by elarlang** [comment]
> I am absolutely on topic here. Please contribute to the conversation positively or hold your comments.

Easy to point fingers, but... Did you read the issue?  Does the defense you propose or address, requires encoding something?

**Comment by jmanico** [comment]
Yes I did read the issue. I do not think you understand how complex this issue is. I am trying to help and would appreciate it if you focus on the positive aspects of what I am trying to say.

**Comment by jmanico** [comment]
Yes.

- If you are placing untrusted data in a URL of a querystring parameter, you need to URL encode so the URL is valid.
- If you are placing untrusted data in the path of a URL REST-style you need to base64url encode to avoid path traversal.
- If you are placing an untrusted URL in an attribute (like src or href), you need to encode in the context of display (typically HTML entity encoding) but you also need to validate that the protocol of the URL is not dangerous since even encoded Javascript url's will still cause XSS.

**Comment by jmanico** [comment]
These two.

- If you are placing untrusted data in the path of a URL REST-style you need to base64url encode to avoid path traversal.
- If you are placing an untrusted URL in an attribute (like src or href), you need to encode in the context of display (typically HTML entity encoding) but you also need to validate that the protocol of the URL is not dangerous since even encoded Javascript url's will still cause XSS.

**Comment by jmanico** [comment]
Please note when placing an untrusted URL in an HTML document, it is one of the only cases when encoding will not stop XSS and is a very special case. It's a common React bypass and is a huge problem from what I see when it comes to googles research. This is why we need a special requirement to discuss this issue which I will add a new issue on later today.

**Comment by elarlang** [comment]
> If you are placing untrusted data in the path of a URL REST-style you need to base64url encode to avoid path traversal.

I'll argue, that you can do it with correct URL encoding. Anyway, it's still encoding and context.

> If you are placing an untrusted URL in an attribute (like src or href), you need to encode in the context of display (typically HTML entity encoding) but you also need to validate that the protocol of the URL is not dangerous since even encoded Javascript url's will still cause XSS.

First part is still encoding, second part is validation and out of scope of this issue.

**Comment by jmanico** [comment]
If you schedule time with me, I will demonstrate to you how if you are URL encoding data that you put in path of a URL it still causes path traverasal in browser and another contexts. I have clear evidence showing that what I am stating is true.

Calendly.com/manicode

This is why the JWT standard uses URL safe encoding coding (base64url) and not URL or other encoding when placing JWT’s in a URL.

**Comment by jmanico** [comment]
> First part is still encoding, second part is validation and out of scope of this issue.

These defenses needs to be done together like we see in Angular. I’ll add a new requirement that merges these defenses. Encoding alone is not helpful or secure when placing URL’s in HTML so the requirement is not helpful from a defensive POV.

**Comment by elarlang** [comment]
It is still encoding and context. 

20+ comments later, I still can not understand, what is your problem statement or proposal for the latest requirement proposal (https://github.com/OWASP/ASVS/issues/1589#issuecomment-2067916530):
> Verify that output encoding is relevant for the interpreter and context required, such as encoding the relevant characters for HTML elements, HTML attribute, JavaScript, URL parameter, HTTP header, or SMTP.

**Comment by jmanico** [comment]
Let me try again.

> Verify that output encoding is relevant for the interpreter and context required, such as encoding the relevant characters for HTML elements, HTML attribute, JavaScript, URL parameter, HTTP header, or SMTP

This requirement only mentioned url parameters and implies url encoding only. The issue around URL’s is much larger and encoding is not a complete defense like it is for other types of data. URL’s are a special case where encoding still leads to path traversal or XSS if you only url encode or skip URL validation. I am happy to give you a few demos to show this to you.

**Comment by elarlang** [comment]
No one ever said that you can only use one requirement to have complete defense.

Can you show me proof of concept, when URL-encoding is not enough for displaying data in any part of the URL?
Can you point to the Google research you mentioned?

**Comment by jmanico** [comment]
I am not saying that the encoding URL requirements are not a complete defense. I am saying that the requirements around URL encoding are both incomplete and incorrect.

The main demo I will email you (a link to my private server) in basically this kind of code in markup.

`<a href="https://somesite.com/data/" + URLEncode(request.getParameter("param"))">link</a>`

I reached out to my contacts at Google and will get you a good article on Javascript URL's ASAP.

**Comment by jmanico** [comment]
This requirement is primarily about contextual encoding. Regardless of the previous comments, the way its worded now it's primarily an XSS defense requirement.

My proposal is that we remove the URL parameter from this requirement. And then we add new requirements that address the complexities of URL handling, validation and encoding that I have addressed above.

If developers apply this requirement as is, it's incorrect and will lead to security vulnerabilities. This is a very complicated issue (URL handling). This basic requirement is absolutely the wrong path IMO and would like someone other than Elar who actually understands the complexities of URL handling to weigh into this conversation, please.

**Comment by jmanico** [comment]
Here is the Google research I mentioned.

https://speakerdeck.com/lweichselbaum/csp-a-successful-mess-between-hardening-and-mitigation

**Comment by elarlang** [comment]
## The scope

The goal for the requirement is written https://github.com/OWASP/ASVS/issues/1589#issuecomment-2068586392

> That it is not possible to manipulate the output command or document structure?

[CWE-116](https://cwe.mitre.org/data/definitions/116.html) "Improper Encoding or Escaping of Output" descibes it also in understandable way:

> The product prepares a structured message for communication with another component, but encoding or escaping of the data is either missing or done incorrectly. As a result, the intended structure of the message is not preserved.


So, it is all about keeping the structure for dynamically built document. I pointed it out, for example here: https://github.com/OWASP/ASVS/issues/1589#issuecomment-2094246129

I also pointed out many times, that the goal for the requirement is not to cover all XSS topic:

https://github.com/OWASP/ASVS/issues/1589#issuecomment-2102746207

> In other words - requirement 5.3.1 is not an "all-in-one defense against every possibility for XSS" but just "You need to use encoding correctly according to the context".

https://github.com/OWASP/ASVS/issues/1589#issuecomment-2102817525

> We are not covering "all the possible XSS here". Already said that, many times. If you have a specific problem of handling URLs. please open a separate issue and stay on topic here.

... and that using another schemes than https just requires input validation, which means that all javascript, data, mailto and other scemes are off-topic here. Including the Google CSP-related presentation.

The requirement is valid for any encoding situation - HTML, JSON, hex, base64, base64uri. Encoding is not only needed against "XSS" attacks in the browser, it is used everywhere, where there is some encoding involved.

If after tens of comments and tens of emails one does not read what is written and keeps pushing their agenda, I call it something seriously disrespectful thing to do. What is the chance to find agreement, if one is talking about one thing and another "fighting against it" with another topic? The chance is 0.


## Path traversal

Path traversal in URL context is valid topic, and defense against that is URL encoding.

The proof-of-concept Jim sent me via email, was carrying the idea that URLs like `https://h2ck3r.eu/legit/../cs.png` work as path traversal even, when partially URL-encoded, (and claims therefore URL encoding is not enough), such as:
```
https://h2ck3r.eu/legit/..%2fcs.png
```

Although it was working on Jim's environment, it does not work for example on the Apache web server. I say, that in Jim's environment, it is incorrect URL handling in place because there is unnecessary URL decoding used and this causes the path traversal issue for the correctly encoded path part.

It is a widespread problem when an application sends part of the URL to an API.


## The "call out"

> This basic requirement is absolutely the wrong path IMO and would like **someone other than Elar who actually understands the complexities of URL handling** to weigh into this conversation, please.

To call me out with sentense, that I don't understand the complexity of URL handling, is just... another "not that nice thing to do". Especially in the context of ignoring the scope.

I'm confident to say that I know a thing or two about encoding, about URL's and about problems with handling URL's. For that exact reason I was passionate to have requirement in ASVS:

> V5.6.1 Verify that input is decoded or unescaped into a canonical form only once and that this is done before processing the input further, for example it is not performed after input validation or sanitization.

Ironically, Jim did not see the need for the requirement (https://github.com/OWASP/ASVS/issues/1590#issuecomment-1937769043).

When making your next move - keep in mind, that in real life, there is no revert or rollback available.

Anything else? Any more examples to analyze?

From email:
> I can show you many other examples of this if needed.

**Comment by elarlang** [comment]
Elar: https://github.com/OWASP/ASVS/issues/1589#issuecomment-2068586392

>> It makes again the requirement longer, but should we mention also the goal for the requirement? That it is not possible to manipulate the output command or document structure?

Josh: https://github.com/OWASP/ASVS/issues/1589#issuecomment-2084885662

> I think we can probably explain the goal in the section text intro. I think trying to include here is a little too complicated.

@tghosth - after the lengthy "discussion" here, I'm more than convinced, we need to mention it in the requirement text.

It also addresses better CRLF injection and whatever other encoding related issue - preserve the message structure.
https://github.com/OWASP/ASVS/issues/1589#issuecomment-1479515632 / https://github.com/OWASP/ASVS/issues/1561

**Comment by jmanico** [comment]
The requirement concerned output encoding, and then someone started talking about CSP, which is a very different issue. But I try my best to embrace all comments. When it comes to encoding, URLs are very special and need their own requirements, so my focus is to fix the "URL parameter" part of this requirement.

**Comment by elarlang** [comment]
With quick recheck I would say, that the first comment on CSP is from you.

And you did not still realize, that your understanding of the URL encoding/decoding is/was wrong? Instead of re-analyzing, testing and re-testing, that your statements are technically correct, you say that I'm the one who can not understand the topic? Doing so, you are surprised to have certain feedback?

**Comment by jmanico** [comment]
I’m very certain that my understanding of URL handling is correct. I even showed you a demo were URL encoding in a path leads to path reversal in certain aspects of the web pipeline. This is why the JWT standard uses base64URL encoding so you can safely place JWT data on a path. There are certain server side URL parsers that have the same problem.

Handling URLs is very challenging and I’ve described it several times already. If you’d like, I’m happy to go through it again with you.

The way that we address URL parameters in this requirement minimizes the complexity of URL based encoding needs.

On a different topic there’s a reason why GO templates will both URL encode a URL query string parsmeter and then attribute encode the full URL if you place it in a src attribute. Sometimes you need to double encode. URL handling is extremely complicated and I don’t want to simplify the issue. I want to explain in requirements why it’s so complicated.

**Comment by elarlang** [comment]
Just wow.

>  I even showed you a demo were URL encoding in a path leads to path reversal in certain aspects of the web pipeline. 

Are you just ignoring the "path traversal" part from my longer comment? I provided you description there, why your demo is NOT correct.

Let me ask it again then, if in an HTML document, there is:
```
<img src="https://domain/one/..%2ftwo%2fthree.png">
```

What is the URL browser is going to ask and what is the file, server is going to serve?

**Comment by jmanico** [comment]
This is just one of many examples. Many server-side URL parsers and processers still allow path traversal when you put untrusted data in a path with URL encoding. It's just an unsafe practice overall. 

My server is not Apache, and I agree the problem is not universal but it iswidespread. This is why the JWT standard went for base64url encoding - for universal defense. 

So yes I read your argument but I do not agree with it just because you cannot get my demo to work on Apache.  It does work elsewhere like I have shown you.

**Comment by elarlang** [comment]
> I do not agree with it just because you cannot get my demo to work on Apache

This is epic facepalm, Jim. You just prove again and again, that you can not understand the need for URI encoding.

If your application unnecessarily decodes the URL and causes path traversal issues because of that, the problem is in your application. The path part was correctly URL-encoded to avoid path traversal.

> This is just one of many examples. Many server-side URL parsers and processers still allow path traversal when you put untrusted data in a path with URL encoding. It's just an unsafe practice overall.

If the server side component causes the problem, it must be fixed there. From this requirement the URL encoding part is done correctly. For that reason I recommended to change in https://github.com/OWASP/ASVS/issues/1590#issuecomment-2103953573

**Comment by elarlang** [comment]
If you can not back up your sentences with proof-of-concepts and fail to admit your mistakes, you should stay away from it.

I ask clear questions and would like to see clear answers, not some bubbles.


## Question 1

Question from: https://github.com/OWASP/ASVS/issues/1589#issuecomment-2111667558

I asked a clear question, but instead of answering it, you did not respond with a clear answer.

If there is a URL:
```
https://domain/one/..%2ftwo%2fthree.png
```

You claim (https://github.com/OWASP/ASVS/issues/1589#issuecomment-2111271913) to certainly know the URL encoding, it should be simple for you. Let the World see your knowledge.

> I’m very certain that my understanding of URL handling is correct.


Anwers options:

  * Option A: Should application serve the file (like you said):
  ```
  https://domain/two/three.png
  ```
  * Option B: Should application serve the file (like I said):
  ```
  https://domain/one/..%2ftwo%2fthree.png
  ```


## Question 2

Quote from: https://github.com/OWASP/ASVS/issues/1589#issuecomment-2102858797

> If you are placing untrusted data in the path of a URL REST-style you need to base64url encode to avoid path traversal.

Do you (now, finally) agree, that saing "URL-encoding path data is not enough against path traversal in the URL" is technically incorrect?

**Comment by elarlang** [comment]
Explanation for others, why I "keep going": In a place like an argument-based discussion for implementing security requirements, there is no place for bluff or demagogy. It causes direct quality issues to the standard and, additionally, kills motivation for other volunteers to waste time with this behavior. So, sorry for the noise, but this is the reason, why I just don't leave it like that - if a person comes and throws sh*t towards me/others, there is a response for that.

**Comment by jmanico** [comment]
The point I am making is that when you url encode data on a path, some servers allow for path traversal like I demonstrated to you on my server. Which is just a HTTP web server (Getty).

This is why the JWT standard decided to standardize on a filesafe standard for its encoding, base64url encoding. Base64url encoding is a universally safe way to add any data, including binary data, to the path of a URL. URL encoding as I demonstrated to you, is not.
 
Thank you for this discussion. I hope my summary helps the cause.

**Comment by jmanico** [comment]
> Do you (now, finally) agree, that saing "URL-encoding path data is not enough against path traversal in the URL" is technically incorrect?

Absolutely not. I already demonstrated to you how url encoding causes path traversal on some servers. URL encoding is absolutely not a universally safe way to add untrusted data to a path. No way.

**Comment by jmanico** [comment]
> Option B: Should application serve the file (like I said):
https://domain/one/..%2ftwo%2fthree.png
The fact is that some servers, some popular servers, will indeed serve this content and url encoding will not stop the path from changing.
This is why as an educator I unfortunately cannot teach that url encoding is a good way to add data to a REST path. I got called out by a student who demonstrated to me how he was still hit with path traversal (I believe) on IIS and other servers.
This is why I changed my course content. I was given clear evidence that url encoding on a path is not a universally safe defense! It was embarrassing, I used to teach url encoding for years! So Elar I am not trying to hurt or offend you. I’m trying to come up with security rules that are universally safe. URL encoding is not universally safe when it comes to this situation.

**Comment by tghosth** [comment]
Gents, I would really like to get this discussion back on track.

Having read through, it seems like we are all in agreement that encoding for a document (such for an HTML response, HTTP response headers, other document generation) needs to be done correctly based on context, e.g. whatever goes into an HTML attribute has to be encoded correctly for an HTML attribute, be it a URL or whatever.

Separately, there is a complicated discussion going on regarding the safe creation of URLs. I would like to separate out this discussion into #1961 as I believe it is complicated enough to merit it's own requirement. Feel free to copy relevant points from here into there. 

@jmanico on that basis, can we proceed with the following wording for this requirement plus the outcome of #1961.

> Verify that output encoding is relevant for the interpreter and context required, such as encoding the relevant characters for HTML elements, HTML attribute, JavaScript, URL parameter or HTTP header, to avoid changing the document structure or functionality.

**Comment by jmanico** [comment]
With respect, I’d like to remove URL parameters from this requirement and handle that in a separate series of requirements around URL handling. 

Depending on what you’re doing with a URL you need to do different kinds of encoding as discussed. And sometimes encoding will not provide safety.

**Comment by elarlang** [comment]
The issue is comment-by-comment analyzed by 2 co-leaders and the proposal is the result of that.

I provide the summary and argumentation for the discussion, although all points are already covered in the comments.

It is important to understand the meaning of the requirement - it is "to avoid changing the document structure or functionality". Whatever is not talking about this goal, it is not responsibility for this requirement and should stay away from this discussion (validating and using the URL by interpretators).

The requirement is also valid for URL parameters or URL paths.

https://datatracker.ietf.org/doc/html/rfc1738

> All unsafe characters must always be encoded within a URL. For
>   example, the character "#" must be encoded within URLs even in
>   systems that do not normally deal with fragment or anchor
>   identifiers, so that if the URL is copied into another system that
>   does use them, it will not be necessary to change the URL encoding.


If we take an URL:
```
<scheme>://<host>:<port>/<path>?<searchpart>
```

If building an URL, any part of it must follow correct encoding according to context, to keep the structure - that it is not possible to change URL meaning with special characters (e.g. `/`, `#`, `?` etc).

It is offtopic from this requirement point of view, but as it seems to cause confusion, I explain that as well - if an interpreter (web server, API) causes a path traversal issue because it unnecessarily URL-decodes the value and causes security risks like path traversal, then there can be a proposal to base64 encode the value and send this instead.

It is a logical value, but the base64 encoded value must be URL encoded, when used in any part of the URL, because it may contain characters, that require URL encoding, such as `/`, `+`, and `=`. For that exact reason, base64uri was invented.

But in all given examples, the proposed requirement stays valid - whatever is the logical value to use as part of the URL, it must be correctly URL-encoded. If there is an extra logical security risk to cover for interpreters (which is again another separate topic), it does not make this requirement incorrect.

Also worth pointing out - this requirement has been there pretty much as it is, we only shortened it a bit and clarified the meaning.

I'm confident to say, that the requirement is technically correct and valid, without taking away the possibility that it will be modified if there are new _valid_ arguments to do so.

**Comment by jmanico** [comment]
There are two forms of base 64 encoding. The first is base64 which is not file/url path safe due to the / and + encoding characters. I am suggesting base64url encoding which is indeed one of the only filepath/url safe standards there is.Base64url encoding is safe because it replaces the / and + with filesafe characters and is specially meant for use on a url path for binary or other data.I mean no disrespect here. I’m only commenting because I saw a base64 encoding mentioned which is not path safe.Base64URL | Base64 Standards | Base64base64.guru

**Comment by jmanico** [comment]
I regret not reading your post to completion and will strive to do so in the future.

**Comment by tghosth** [comment]
The continued discussion above reinforces for me that we need to move discussion of URLs to a separate requirement as per #1961.

I am undecided on URL parameter and I might leave this to our next meeting to discuss.

**Comment by tghosth** [comment]
The agreed wording is as follows but we are waiting for the outcome of #1961:

> Verify that output encoding is relevant for the interpreter and context required, such as encoding the relevant characters for HTML elements, HTML attribute, JavaScript, or HTTP header, to avoid changing the document structure.

**Comment by elarlang** [comment]
To be clear with that - it **was not agreed** from my side, that URL parameters should be removed from the requirement. Conditional "waiting for #1961" means that it should provide a duplicate to this requirement to cover URL encoding only. 

This comment is an add-on to my previous comment: https://github.com/OWASP/ASVS/issues/1589#issuecomment-2116829319

**Encoding** - It is needed against _injection_ attacks, if the document is built dynamically, user input is placed between some limiters, for keeping the document structure.

If someone ever reads this issue, here are my arguments. I put numbers to points, then it is easier for me to say, which point was ignored (again) when (not) reading my arguments.


## Point 1 - Injection vs Execution.

Injection - when an attacker can change the dynamically built document structure

* program-code: `<a href="` + userinput + `">link</a>`
* userinput:
  * `"><script>evil-code</script>` - HTML injection
  * `" onclick="evil-code" style="over-the-screen` - HTML attribute injection

Execution - when an attacker can execute malicious code, without manipulating the document structure during building it - so **encoding does not help in this situation**

* program-code: `<a href="` + userinput + `">link</a>`
* userinput:
  * `javascript:evil-code` - JavaScript execution

Encoding the user input here is not enough, as it is logical problem - untrusted input reached to HTML attribute and the problem here is, it was not validated.

Again, this example is input validation problem, and you can not fix it with any encoding. To say, that "URL-encoding is not enough because of javascript:" is just _filtered_. The responsibility for encoding is to keep the document structure. The responsibility for validation is to avoid untrusted data to reach to content execution.

For JavaScript code, execution attacks are often caused because of `eval`, applying untrusted data to `innerHTML`, `document.write()` etc. Although (double)encoding may help there, by principle it is an incorrect defense.

Content execution is also the situation when HTML and JavaScript get executed because of an incorrect `Content-Type` header, let's say `text/html` instead of `application/json`. Instead of encoding the data, you need to just set the correct `Content-Type`.


## Point 2 - It is not limited to HTML syntax or document

The requirement is as principle, it is valid for every syntax and for every context when user input is used for dynamically built documents and is placed between delimiters.

JavaScript:
```js
var x = 'USERINPUT';
```

You can use encoding to unicode, e.g. `'` is `\u0027` (and other alternatives). In this case one option is escaping with the same goal - to keep the document structure.

JSON (== JavaScript), no one should build it manually:
```js
{"x":"USERINPUT"}
```

The same, as described with JavaScript.

URL-encoding when building a URL:
```
scheme://hostname/USERINPUT1/?x=USERINPUT2#USERINPUT3
```

In any part of the URL, you need to use URL encoding to be sure it does not change the structure of the URL.

If the value is something, that is "URL-safe" (base64uri encoded value), that does not contain any characters that require URL-encoding, then URL-encoding does not make any changes to the value, but, the URL-encoding should be there anyway.

Without URL-encoding this URL-safe part, your defense is relying on the expectation, that this part is always URL-safe, but it is a business logic rule. What if there is a need to send something else to that place that is not URL-safe anymore?


HTTP header, classical mistake:
```
Content-Disposition: attachment; filename="USERINPUT"
```

Note, that expected defense in this case is to sanitize the input for the `filename` attribute and encode it correctly for `filename*` attribute. The point is, encoding is not an HTML document specific problem.


You can add here CSS, XML, "any plain-text syntax", ...


## Point 3 - It is not limited to XSS

However many times I wrote the goal for the requirement is not avoiding all possible XSS vectors, it's just keeping the document structure in place. It is just an important part of the defense against XSS if certain technologies or syntaxes are involved.


## Point 4 - Limiters!

Often encoding helps **only** when user input goes between some limiter marks, and the point for the encoding is - you should stay between those limiter marks.

* In HTML document context - `>` and `<`
* In HTML attribute value context - `'` or `"`
  * Note that, if user input is placed without delimiter marks, encoding does not help, `<input name=x value=USERINPU>`
* JavaScript - `'`, `"`, or template markers `\``

Examples, when encoding does not help:

* HTML elemenent attriburte name `<input name=x USERINPUT>`
* JavaScript: `var x = USERINPUT;`
* JSON `{USERINPUT:"value"}`

In those cases, not validated (and if needed, then also sanitized) user input must never reach to this functionality.


So encoding is just one tool with a really precise goal - to keep the structure for dynamically built documents. Whatever is outside of this goal, is out of scope for this requirement AND does not make this requirement incorrect, **including the "URL parameters"**.

**Comment by jmanico** [comment]
You make a very solid argument. You were outvoted by two other ASVS leads (for now) because:

1) Even when encoded, JavaScript: URL’s are still dangerous in a webpage. This is a very special case where encoding does not fully provide safety. It’s in fact the only case.

2) Many web servers have issues with URL encoding in a path that still allows for path traversal.

3) URL safety is very necessary for defense when it comes to the risks of XSS, open redirect, SSRF and sometimes access control defense.

Encoding, dynamically building and validating URL’s is extremely challenging for developers and relates to a wide variety of AppSec risk.  This is why they deserve special treatment.

**Comment by elarlang** [comment]
> You make a very solid argument. You were outvoted by two other ASVS leads (for now) because:

By concept, if "voting" goes over technically provided arguments, it's the dead-end.

1) comment is out of scope - the scope is to keep the structure.
2) comment is out of scope - web server handles URL as interpreter, it is another topic
3) comment is out of scope - comment is also against / in conflict the point made
4) each defense method has its own place in the flow, encoding does not replace validation, and validation does not replace encoding. Do not mix them.

In conclusion, it does not matter at all, what arguments I provide, you still come "against" the points with things, that I just described to be out of scope.

**Comment by tghosth** [comment]
From my perspective, the removal of "URL parameter" relates to the complexity of handling URLs which I believe should be included elsewhere.

I think there is still a question here related to document structure which we should discuss further.

**Comment by jmanico** [comment]
I’ve been thinking a lot about this and am trying to reduce my concerns to the bare minimum.

Base64url encoding is not that important to safe document construction. It’s a little important when adding binary data or JWT’s to a url.

But validating untrusted URL’s is crucial to safe document construction. URL’s are the only case where encoding will still lead to client side injection and that is such a common successful attack I want it to be addressed a lot more prominently when it comes to document construction.

So my bottom line is:

1) give url validation a very prominent requirement when it comes to document construction. Mentioning url encoding by itself without mentioning validation is dangerous.
2) somehow directly mention base64url encoding when it comes to adding binary data or JWT’s to a url

I do not think the current requirements address these prominently enough.

**Comment by tghosth** [comment]
> I think there is still a question here related to document structure which we should discuss further.

Having discussed this further with Elar, the direction becomes a little too general because when you start worrying about building HTML attribute names or element names dynamically or building JSON attribute names dynamically you are getting into the territory of allowing the end user to write their own code as well :) This can't be fixed by encoding so we will leave it for now.


This returns us to how to handle this requirement.

My current feeling is that the fact that the current 5.3.1 requirement is a little too abstract is one of the things that has led to this long discussion.

My current feeling is to have specific requirements which cover the key cases and then a catch-all output encoding for other cases which describes a function's behaviour "takes user input and builds some other representation or command based upon it".

Possible rough splitting for now:

> Verify that output encoding for an HTTP response/HTML document/XML document is relevant for the context required, such as encoding the relevant characters for HTML elements, HTML attribute, JavaScript, or HTTP headers, to avoid changing the message or document structure.

> Verify that the correct encoding is performed for each context when generating a URL.

> Verify that the correct encoding is performed for each context in a JSON document.

**Comment by tghosth** [comment]
@tghosth to draft some requirements that sit in between 1 requirement for all of the above and individual requirements for everyone one of the above.

**Comment by tghosth** [comment]
| Handled | Category | Requirement | Notes|
|:---:|---|---|---|
| ✅ | HTML Docs | Verify that user-generated content in HTML tags is properly encoded. | Covered by [5.3.1](https://appsecg.host/5.3). |
| ✅ | HTML Docs | Verify that user-generated content in HTML attributes is properly encoded. | Covered by [5.3.1](https://appsecg.host/5.3). |
| ❌ | HTML Docs | Verify that data embedded in JavaScript is encoded using a safe method like JSON encoding or JavaScript-specific encoding functions. |  |
| ❔ | HTML Docs | Verify that data in CSS contexts is encoded to avoid CSS Injection. | Mentioned in both [5.3.1](https://appsecg.host/5.3) and [5.2.8](https://appsecg.host/5.2) |
| ✅ | HTML Docs | Verify that data within JSON documents is properly encoded to prevent injection attacks. | Mentioned in [5.3.6](https://appsecg.host/5.3) and discussed in #1555 |
| ✅ | HTML Docs | Verify that the correct encoding is performed for each context when generating a URL. | Being discussed in #1961 |
| ✅ | HTML Docs | Verify that HTTP headers generated from user input are properly validated and encoded. | Covered by [5.3.1](https://appsecg.host/5.3). |
| ❌ | HTML Docs | Verify that dynamically generated HTML is constructed safely using templates or frameworks that auto-encode outputs. |  |
| ❔ | HTML Docs | Verify that HTML comments do not contain untrusted user input. | Should add to [5.3.1](https://appsecg.host/5.3). |
| ✅ | Email | Verify that email content and headers do not include untrusted user input without proper encoding and validation. | Covered by [5.2.3](https://appsecg.host/5.2). |
| ✅ | LDAP | Verify that LDAP queries are built using safe methods that separate code from data. | Covered by [5.3.7](https://appsecg.host/5.3) |
| ✅ | Database | Verify that SQL queries use parameterized queries or prepared statements. | Covered by [5.3.4](https://appsecg.host/5.3). |
| ✅ | Database | Verify that ORM (Object-Relational Mapping) frameworks are used correctly to prevent SQL Injection. | Covered by [5.3.4](https://appsecg.host/5.3). |
| ✅ | Database | Verify that stored procedures are designed to accept parameters that use parameterized queries or prepared statements. | Opened #1993 to include this in [5.3.4](https://appsecg.host/5.3). |
| ✅ | XML | Verify that XML documents use proper encoding to avoid XML Injection. | Covered by [5.3.10](https://appsecg.host/5.3). |
| ✅ | XML | Verify that input used in XML parsing is validated and entity expansion is disabled. | XXE is covered in [5.3.10](https://appsecg.host/5.3) and schema validation is in [13.3.1](https://appsecg.host/13.3). |
| ✅ | OS | Verify that command-line arguments are properly escaped or use safe API functions. | Covered in [5.3.8](https://appsecg.host/5.3). |
| ✅ | OS | Verify that input used in command execution is properly validated and sanitized. | Covered in [5.3.8](https://appsecg.host/5.3). |
| ✅ | Logs | Verify that input used in log messages is properly encoded and sanitized. | Covered in [7.3.1](https://appsecg.host/7.3). |

**Comment by jmanico** [comment]
> @jmanico do you think this is sufficiently covered in [5.2.5](https://appsecg.host/5.2)?

Politely, no. I think safely building HTML (encoding and similar) and avoiding template injection (avoid server side template assembly) are different controls.

**Comment by jmanico** [comment]
> "Verify that data embedded in JavaScript is encoded using a safe method like JSON encoding or JavaScript-specific encoding functions."	@jmanico do you think this is sufficiently covered in [5.3.1](https://appsecg.host/5.3)?

Yes, for sure.

**Comment by jmanico** [comment]
> I think we need both 5.2.8 and 5.3.1
> 5.2.8 | Verify that the application sanitizes, disables, or sandboxes user-supplied scriptable or expression template language content, such as Markdown, CSS or XSL stylesheets, BBCode, or similar.
-- | --

This addresses sanitizing user authored cunks of CSS. Just like untrusted HTML is need to be sanitized.

> 5.3.1 | [MODIFIED] Verify that output encoding is relevant for the interpreter and context required. For example, use encoders specifically for HTML values, HTML attributes, JavaScript, CSS, URL parameters, HTTP headers, SMTP, and others as the context requires, especially from untrusted inputs (e.g. names with Unicode or apostrophes, such as ねこ or O'Hara).
-- | --

This addresses adding user controlled data, like a color, into blocks of otherwise static css. Different controls and to some degree different attacks, too.

**Comment by tghosth** [comment]
I updated the table above.

**Comment by tghosth** [comment]
I merged this in. I added a catch all for any encoding scenarios not covered elsewhere [as I suggested here](https://github.com/OWASP/ASVS/issues/1589#issuecomment-2152050611).

I don't think this wording is perfect but it has taken too long to spend further on it for now.

**Comment by elarlang** [comment]
> I merged this in. I added a catch all for any encoding scenarios not covered elsewhere https://github.com/OWASP/ASVS/issues/1589#issuecomment-2152050611.

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **5.3.1** | [MODIFIED] Verify that output encoding for an HTTP response/HTML document/XML document is relevant for the context required, such as encoding the relevant characters for HTML elements, HTML attributes, HTML comments, JavaScript, CSS, or HTTP headers, to avoid changing the message or document structure. | ✓ | ✓ | ✓ | 116 |
| **5.3.14** | [ADDED] Verify that output encoding is relevant for the interpreter and context required in any context where a potentially dangerous interpreter, not mentioned above, is being used. | | ✓ | ✓ | |

I prefer to not have the 5.3.14 and have 5.3.1 to cover all cases.

The requirement text must be independent and understandable when watched just a requirement alone - it can not contain "not mentioned above".

**Comment by tghosth** [comment]
I'd really rather keep 5.3.1 focused on it's particular area. Would it be better to just change the text to say:

> [ADDED] Verify that output encoding is relevant for the interpreter and context required in any context where a potentially dangerous interpreter ~, not mentioned above,~ is being used.

**Comment by jmanico** [comment]
Looks good @tghosth @elarlang

**Comment by tghosth** [comment]
Whilst the concept of a catch-all makes sense, having thought about it more, I don't think it adds that much value compared to adding another, potentially confusing requirement.

**Comment by tghosth** [comment]
We also have 5.2.2 as a sort of catch all for dangerous sinks🙃

**Comment by tghosth** [comment]
@set-reminder 7 days @tghosth to remove 5.3.14

**Comment by tghosth** [comment]
5.3.14 should be removed in #2026
